test/StateSchedulingTest.cpp

```diff
@@ -0,0 +1,125 @@
+#include "gtest/gtest.h"
+#include <iostream>
+#include <unistd.h>
+
+namespace {
+
+class Base {
+ public:
+  int GetValue() const { return value_; }
+ protected:
+  int value_ = 0;
+};
+
+class Processor : public Base {
+ public:
+  void Process(::pipert::PacketToProcess<int> packet) {
+    for (int i = 0; i < 10; ++i) {
+      value_ += packet.data();
+    }
+    for (int i = 0; i < 10; ++i) {
+      value_ -= packet.data();
+    }
+  }
+};
+
+class Incrementor : public Base {
+ public:
+  Incrementor(::pipert::ScheduledChannel<int>&& decChannel) : decChannel_(std::move(decChannel)) {}
+  void Increment(::pipert::PacketToProcess<int> packet) {
+    value_ += packet.data();
+    ::pipert::PacketToFill<int> new_packet = decChannel_.Acquire("Decrementor", packet.timestamp(), packet.data());
+    while(packet.IsEmpty()) {
+      new_packet = decChannel_.Acquire("Decrementor", packet.timestamp(), packet.data());
+    }
+    new_packet.Push();
+  }
+
+  int GetDroppedPacketsNumber() {
+    return decChannel_.GetDroppedPacketsNumber();
+  }
+ private:
+  ::pipert::ScheduledChannel<int> decChannel_;
+};
+
+class Decrementor : public Base {
+ public:
+  void Decrement(::pipert::PacketToProcess<int> packet) {
+    value_ -= packet.data();
+  }
+};
+
+}  // namespace
+
+TEST(StateScheduling, SchedulingWithoutState) {
+  ::pipert::Scheduler sch(4);
+  Processor processor;
+  ::pipert::ScheduledChannel<int> incrementer = sch.CreateScheduledChannel<int>("Incrementer", 4, nullptr, std::bind(&Processor::Process, &processor, std::placeholders::_1));
+
+  sch.Start();
+  for (int i = 0; i < 100; ++i) {
+    ::pipert::PacketToFill<int> packet = incrementer.Acquire("Incrementer", ::pipert::Timer::time(), 1);
+  }
+   
+
+  sch.Stop();
+  std::cout << processor.GetValue() << std::endl;
+  EXPECT_TRUE(-1000 <= processor.GetValue() && processor.GetValue() <= 1000);
+}
+
+TEST(StateScheduling, SchedulingWithState) {
+  ::pipert::Scheduler sch(4);
+  Processor processor;
+  ::pipert::ScheduledChannel<int> incrementer = sch.CreateScheduledChannel<int>("Processor", 4, &processor, std::bind(&Processor::Process, &processor, std::placeholders::_1));
+
+  sch.Start();
+  for (int i = 0; i < 100; ++i) {
+    ::pipert::PacketToFill<int> packet = incrementer.Acquire("Processor", ::pipert::Timer::time(), 1);
+  }
+   
+  sch.Stop();
+  EXPECT_EQ(processor.GetValue(), 0);
+}
+
+TEST(StateScheduling, SchedulingDropCheck) {
+  ::pipert::Scheduler sch(4);
+  Decrementor decrementor;
+  ::pipert::ScheduledChannel<int> decChannel = sch.CreateScheduledChannel<int>("Decrementor", 4, &decrementor, std::bind(&Decrementor::Decrement, &decrementor, std::placeholders::_1));
+  sch.Start();
+  for (int i = 0; i < 100; ++i) {
+    ::pipert::PacketToFill<int> packet = decChannel.Acquire("Decrementor", ::pipert::Timer::time(), 1);
+    while (packet.IsEmpty()) {
+      packet = decChannel.Acquire("Decrementor", ::pipert::Timer::time(), 1);
+    }
+    packet.Push();
+  }
+  usleep(1000);
```

Comment:
I don't know what could be a consistent solution, or how could we know when all subprocesses are finished?

Reply:
Right now, it seems like there might not be any consistent  way to tell when processing of a pipeline is completed. That's probably because when the scheduler is running, we suppose there might always be new packages pushed into the start of the pipeline, so processing is never truly complete, in a sense.

But, if we know here that we won't push anything else onto the pipeline, such as here, we should probably give some way to tell outside of the scheduler whether it's currently working on anything or not. I think the only way to tell if a pipeline has completed processing is to check each stage (channel) to see that there are no queued packets for processing. Maybe we could make a new function (slightly expensive to call often) that checks this and tells us if there are any packets in transit in the pipeline. This may be the subject for a new issue?

Reply:
Yes, there is no finish line as the pipeline is always ready to accept more input data. But if you call stop(), it will immediately instruct the threads to shut down although it will wait for them to do so. So waiting a little is a good idea, or even better, yielding and checking again will make the test end sooner.

Please use std::something::sleep() instead of using unistd.h as it is a POSIX thing (and may not work on... the something we don't call on its name).